### ISSUE TITLE
Further Speed up FA3 Backend

### ISSUE THREAD

**Issue by hebiao064** [sgl-project/sglang#5810]
We explored and discussed some ideas and we want to write it down for tracking, also welcome community developer to try out those unfinished

- [ ] GQA head packing.
- [x] Split-KV. aka Flash Decoding: We already enabled it, it is indeed faster in lower batch and long context scenario.
- [ ] PDL: https://github.com/Dao-AILab/flash-attention/commit/000090d02f0398e9087a8823fc1f5242becfac99
- [ ] Prepare Scheduler Metadata: https://github.com/Dao-AILab/flash-attention/commit/fa60e7cc97300b4b26721983df580a7da7a8ebea
- [ ] Skip `len` operation, get it directly from forward batch: https://github.com/hebiao064/sglang/blob/main/python/sglang/srt/layers/attention/flashattention_backend.py#L333

**Comment by sanchit-ahuja** [comment]
@hebiao064 Can I pick this up?

**Comment by hebiao064** [comment]
@sanchit-ahuja sure, please let me know which part you want to pick up

**Comment by sanchit-ahuja** [comment]
I can pick up the first and second to-do items for now and can look into PDL and the llama spec decoding issue later. Also, what benchmark do we need to run for the second item?

**Comment by hebiao064** [comment]
@sanchit-ahuja https://github.com/sgl-project/sglang/pull/5611#issuecomment-2819965621
Refer to this one, but you don't need to run it against Deepseek so you don't need to compile DeepGemm first.

**Comment by hebiao064** [comment]
Another potential issue for CUDA Graph: https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/layers/attention/flashattention_backend.py#L1590

need to change to 
```bash
metadata.cu_seqlens_k.copy_(
                torch.nn.functional.pad(
                    torch.cumsum(seq_lens, dim=0, dtype=torch.int32), (1, 0)
                )
            )
```

to ensure CUDA Graph stability

**Comment by sanchit-ahuja** [comment]
Got it. Thanks!

**Comment by sanchit-ahuja** [comment]
@hebiao064 Can you let me know what do we need to do for PDL and how you want me to run profiling for Lllama models?

**Comment by sanchit-ahuja** [comment]
Also, I was looking into the `pack_gqa` issue and found out that this flag is part of the `sgl-kernel` within the repo. How can I expose this flag to the sglang framework for me to run the benchmark? There's also a `test_flash_attention.py` file within the the `sgl-kernel`'s `tests` folder which disables `pack_gqa` test explicitly with a `DISABLE_PACKGQA` set to True. Do you want me to run that as well?

**Comment by sanchit-ahuja** [comment]
I did try running the test `test_flash_attention.py` but I think my GPU is not supported to run this. I have an A100 machine and I am getting this error: `flash_attn at sgl-kernel is only supported on sm90 and above`. Let me know how you want me to proceed ahead with this. Although, FA3 does support A100 now [ref](https://github.com/Dao-AILab/flash-attention/issues/1481).
Thanks!

**Comment by hebiao064** [comment]
> I did try running the test `test_flash_attention.py` but I think my GPU is not supported to run this. I have an A100 machine and I am getting this error: `flash_attn at sgl-kernel is only supported on sm90 and above`. Let me know how you want me to proceed ahead with this. Although, FA3 does support A100 now [ref](https://github.com/Dao-AILab/flash-attention/issues/1481). Thanks!

https://github.com/sgl-project/sglang/issues/5536#issuecomment-2837609533

**Comment by sanchit-ahuja** [comment]
>(Good first issue) Skip len operation, get it directly from forward batch: https://github.com/hebiao064/sglang/blob/main/python/sglang/srt/layers/attention/flashattention_backend.py#L341 , https://github.com/hebiao064/sglang/blob/main/python/sglang/srt/model_executor/forward_batch_info.py#L144

> Another potential issue for CUDA Graph

These are done.

Do you want me to raise a PR for these fixes? In the meantime, I can wait for a fix for the Ampere GPUs. Will see if I can get access to an H100 machine somehow.

**Comment by hebiao064** [comment]
New Thread: 
- [ ] Learn why Triton Backend is fast for Llama Normal decoding


FA3 with CUDA Graph:
- Prepare and invoke Cuda graph: 1ms 213us 801ns
- cudaDeviceSynchronize: 6ms 61us 903ns

Triton  with CUDA Graph:
- Prepare and invoke Cuda graph: 1ms 53us 548ns
- cudaDeviceSynchronize: 5ms 840us 839ns

FA3 without CUDA Graph:
- forward: 21ms

Triton without CUDA Graph:
- forward: 22ms


An interesting finding is that FA3's init metadata replay is faster than triton backend even tho triton backend did it in triton kernel.

236us
![Image](https://github.com/user-attachments/assets/e79b719f-24a2-4797-8710-16b9843a5000)

267us
![Image](https://github.com/user-attachments/assets/c058b52d-8f25-440e-ada1-68a12dab7096)

**Comment by hebiao064** [comment]
> > (Good first issue) Skip len operation, get it directly from forward batch: https://github.com/hebiao064/sglang/blob/main/python/sglang/srt/layers/attention/flashattention_backend.py#L341 , https://github.com/hebiao064/sglang/blob/main/python/sglang/srt/model_executor/forward_batch_info.py#L144
> 
> > Another potential issue for CUDA Graph
> 
> These are done.
> 
> Do you want me to raise a PR for these fixes? In the meantime, I can wait for a fix for the Ampere GPUs. Will see if I can get access to an H100 machine somehow.

Thanks a lot for your interest! But we definitely need H100 to test the PR end to end, it's okay, I can find someone else to fix that or I will fix it when i got time.
I'll keep an eye for you to find suitable task with Ampere GPUs, @yinfan98 Hi Fan, please let sanchit know if there are any good first issue about Ampere testing or optimization
